### DISCUSSION TITLE
Error with new Force Field

### DISCUSSION THREAD

**Discussion by rinkyimrk** [marrink-lab/vermouth-martinize#641] [Q&A] [answered]
I am trying to construct a protein model with a new force field downloaded from the following repository:
https://github.com/KULL-Centre/_2023_Thomasen_Martini/tree/main/force_field

I attempted to generate a topology file using the following command:

`martinize2 -f t4l_clean.pdb -o t4l.top -x t4l_cg.pdb -ff-dir martini_v3.0.0_proteins/force_fields/ -map-dir martini_v3.0.0_proteins/mappings/`

However, I encountered the following error message:

```
Traceback (most recent call last):
  File "../MARTINI/.martini_env/lib/python3.12/site-packages/vermouth/parser_utils.py", line 239, in parse_section
    return method(self, line, lineno, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "../MARTINI/.martini_env/lib/python3.12/site-packages/vermouth/map_parser.py", line 718, in _blocks
    block = getattr(self.force_fields[self.ff[direction]], map_type+'s')[attrs['resname']]
                    ~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^
KeyError: 'universal'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "../MARTINI/.martini_env/lib/python3.12/site-packages/vermouth/map_input.py", line 435, in read_mapping_directory
    new_mappings = read_mapping_file(infile, force_fields)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "../MARTINI/.martini_env/lib/python3.12/site-packages/vermouth/map_input.py", line 178, in read_mapping_file
    for mapping in mappings:
                   ^^^^^^^^
  File "../MARTINI/.martini_env/lib/python3.12/site-packages/vermouth/parser_utils.py", line 110, in parse
    result = self.dispatch(line)(line, lineno)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "../MARTINI/.martini_env/lib/python3.12/site-packages/vermouth/parser_utils.py", line 241, in parse_section
    raise IOError("Problems parsing line {}. I think it should be a "
OSError: Problems parsing line 8. I think it should be a '['modification', 'from blocks']' line, but I can't parse it as such.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "../MARTINI/.martini_env/bin/martinize2", line 1114, in <module>
    entry()
  File "../MARTINI/.martini_env/bin/martinize2", line 755, in entry
    partial_mapping = read_mapping_directory(directory, known_force_fields)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "../MARTINI/.martini_env/lib/python3.12/site-packages/vermouth/map_input.py", line 437, in read_mapping_directory
    raise IOError('An error occured while reading "{}".'.format(path))
OSError: An error occured while reading "../MARTINI/tia1/martini_v300_pps/mappings/modifications.mapping".
```

I suspect there may be an issue with how I’ve set up the mappings or force field directories, but I’m not sure how to resolve it. Has anyone encountered this issue or have advice on troubleshooting? Any guidance would be appreciated.
Thank you.

**Comment by pckroon** [comment, marked as answer]
The "universal" forcefield has been retired a few releases back, and renamed to "charmm". You'll need to adapt the mappings to reflect this change. In the `[from]` section of the mapping files it should read "charmm", instead of "universal". You can probably do something like `find . -iname '*\.map*' -exec sed -i -re /universal/charmm/g {}+` to get it all in one batch, but check my work ;)

**Reply by rinkyimrk** [reply]
Thank you for your response!

After following your suggestion, the first error was solved. However, I encountered another error. Upon investigation, I found that the issue was caused by incorrect atom names in the [ mapping ] sections for N-ter and NH2-ter.

```
[ from blocks ]  
N-ter  
[ to blocks ]  
N-ter  

[ mapping ]  
HN2 BB  <-- Originally included HN1  
HN3 BB  
N   BB  
CA  BB  
C   BB  
O   BB  
```

After making these changes, I ran the following command:

`martinize2 -f t4l_clean.pdb -o t4l.top -x t4l_cg.pdb -ff-dir martini_v300_pps/force_fields/ -map-dir martini_v300_pps/mappings/  `
However, the resulting molecule_0.itp still did not behave as expected. Specifically, I noticed that _PRO was supposed to be added to the entries, but it was missing. The [ atoms ] section looks like this:

```
[ atoms ]  
  1 Q5     1 MET BB    1    1     
  2 C6     1 MET SC1   2  0.0     
  3 P2     2 GLU BB    3  0.0  
```

I suspect these atom names are not as expected, as they may all require the _PRO suffix.

**Reply by pckroon** [reply]
Oops, I forgot to mention that some atoms might have changed name. Happy you managed to figure that out.

I'm not intimately familiar with the forcefield you're using, but it does look like something is misbehaving. I would suggest renaming the custom forcefield to something other than `martini3001`, to make sure it doesn't overlap/collide with the built-in martini3 ff. 
It should be a matter of renaming the ff folder (to, e.g., martini300pps), and adjusting the mappings accordingly (`find . -iname '*\.map*' -exec sed -i -re /martini3001/martini300pps/g {}+`. Or something.). Then you'd need to adjust your CLI invocation to add a `-ff martini300pps`

**Reply by rinkyimrk** [reply]
Thank you for your reply. I’m happy to share that the issue has been resolved.

**Reply by pckroon** [reply]
Awesome. For posterity, was the solution indeed renaming the force field?

**Reply by rinkyimrk** [reply]
Sorry for not providing details. Actually, renaming the force field was not necessary. After switching from "universal" to "charmm" and removing HN1, it worked. The subsequent error was due to my mistake—it was a directory structure issue. Using the directory structure as it was downloaded solved the problem.
